### PR TITLE
SAK-42959 Default lang-datepicker to user's local timestamp

### PR DESCRIPTION
**Pull request by adrianfish**
https://jira.sakaiproject.org/browse/SAK-42959

### CODE REVIEW

**Comment by steveswinsburg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sakaiproject/sakai/pullRequest?prid=5027536&bid=16567958)

File: portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/handlers/SiteHandler.java [hunk 1]
```diff
@@ -556,6 +556,10 @@ public void doSite(HttpServletRequest req, HttpServletResponse res, Session sess
 		rcontext.put("showFavStarsOnAllFavSites",ServerConfigurationService.getBoolean(SAK_PROP_SHOW_FAV_STARS_ON_ALL, SAK_PROP_SHOW_FAV_STARS_ON_ALL_DFLT));
 		
 		addLocale(rcontext, site, session.getUserId());
+
+		addServerTime(rcontext);
+
+		addUserTimezoneOffset(rcontext);
```
Comment (ottenhoff) [inline portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/handlers/SiteHandler.java]: Just one method called addUserTimezoneInfo ? And then we add 3-4 new vars around time?

File: portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/handlers/BasePortalHandler.java [hunk 2]
```diff
@@ -209,4 +210,14 @@ protected void addLocale(PortalRenderContext rcontext, Site site, String userId)
...
+
+	protected void addServerTime(PortalRenderContext rcontext) {
+		rcontext.put("serverTime", Instant.now().toEpochMilli());
+	}
+
+	protected void addUserTimezoneOffset(PortalRenderContext rcontext) {
+
+		int offset = timeService.getLocalTimeZone().getOffset(Instant.now().toEpochMilli());
+		rcontext.put("userTimezoneOffset", offset);
+	}
```
Comment (ottenhoff) [inline portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/handlers/BasePortalHandler.java]: while we are here, why don't we add the user's current (localized) date/time as ISO-8601 .... and might as well add their preferred timezone (e.g., Europe/Stockholm) .......

File: library/src/webapp/js/lang-datepicker/lang-datepicker.js [hunk 3]
```diff
@@ -1684,40 +1684,47 @@ if(c&&c._defaults.timeOnly&&b.input.val()!==b.lastVal)try{$.datepicker._updateDa
...
-			if (!(initVal == "" && options.allowEmptyDate)){
-				if (!initVal) initVal = new Date();
-			}
+      // finally trim the initVal and make sure it is set so we don't Dec 1969 the user
+      initVal = jQuery.trim(initVal);
 
+      if (!(initVal == "" && options.allowEmptyDate)){
+        if (!initVal) {
+          if (portal.serverTime && portal.userTimezoneOffset) {
+            initVal = moment.utc(parseInt(portal.serverTime)).add(portal.userTimezoneOffset/1000, "seconds").toDate();
```
Comment (ottenhoff) [inline library/src/webapp/js/lang-datepicker/lang-datepicker.js]: you GMT developers live in an offset-free world!

```
 var osTzOffset = (new Date()).getTimezoneOffset();
            initVal = moment(parseInt(portal.serverTime))
              .add(portal.userTimezoneOffset, 'ms')
              .add(osTzOffset, 'm')
              .toDate();
```

**Comment by steveswinsburg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sakaiproject/sakai/pullRequest?prid=5027536&bid=16567958)

**Comment by steveswinsburg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sakaiproject/sakai/pullRequest?prid=5027536&bid=16567958)

**Comment by ottenhoff** [comment]
This is working well in my testing just need NOW button to follow same path

**Comment by adrianfish** [comment]
Making progress

**Pull request merged**